src/Compose/ConcurrentTransitionManagerContainer.cs

```
@@ -1,88 +1,88 @@
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					restoredManagers.Add(Restore(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!restoredManagers.IsEmpty)
-				if (restoredManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Restore(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Restore").Invoke(null, new[] { manager });
-			return manager;
-		}
-
-		public void Snapshot()
-		{
-			foreach (var kvp in Managers)
-				Snapshot(kvp.Key, kvp.Value);
-		}
-
-		private void Snapshot(Type serviceType, ConcurrentBag<object> managers)
-		{
-			var snapshottedManagers = new ConcurrentBag<object>();
-			object manager = null;
-			while (!managers.IsEmpty)
-				if (managers.TryTake(out manager))
-					snapshottedManagers.Add(Snapshot(Helper.MakeGenericType(serviceType).GetTypeInfo(), manager));
-			while (!snapshottedManagers.IsEmpty)
-				if (snapshottedManagers.TryTake(out manager))
-					managers.Add(manager);
-		}
-
-		private object Snapshot(TypeInfo helper, object manager)
-		{
-			helper.DeclaredMethods.Single(x => x.Name == "Snapshot").Invoke(null, new[] { manager });
-			return manager;
-		}
-	}
-}
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Reflection;
+
+namespace Compose
+{
+	internal sealed class ConcurrentTransitionManagerContainer : TransitionManagerContainer
+	{
+		private readonly ConcurrentDictionary<Type, ConcurrentBag<object>> Managers
+			= new ConcurrentDictionary<Type, ConcurrentBag<object>>();
+		private static readonly Type Helper = typeof(GenericHelper<>);
+
+		private static class GenericHelper<T>
+		{
+			private static readonly TypeInfo Disposable = typeof(IDisposable).GetTypeInfo();
+			public static void Restore(object untypedManager)
+			{
+				var manager = untypedManager as DynamicRegister<T>;
+				if (Disposable.IsAssignableFrom(manager.CurrentService.GetType().GetTypeInfo()))
```

In the ANE case though, assuming that the public API has its validation for the arguments, this doesn't necessarily prevent the ONRE at a lower internal level - it shouldn't happen - we're the ones writing the implementation after all - and realistically should be caught by tests (where applicable), but we all know it happens sometimes.

Your comments have made me consider if maybe I'm being/thinking too defensively even for internal implementations though? 

However I still like the fail early option rather than not - at this point in time at least.

Definitely agree with consequences for ONRE's though, they are horrible! :smile: